### DISCUSSION TITLE
Semester 4 Week 4

### DISCUSSION THREAD

**Discussion by BenjaminB109** [SE-TINF22B6/Underwatch#210] [Announcements]
## Announcement: Underwatch Project Testing

In this weekly blog, we will provide an update on the current status of testing for our Underwatch application. The whole testing report is [here](https://github.com/SE-TINF22B6/Underwatch/blob/main/documentation/Test%20Report.md) .

The project is structured into three components:
- **Game (major)**
- **Web Page (major)**
- **Backend (minor)**

### Testing Overview

**Backend Tests**  
We utilize SpringBootRest for built-in edge case and functionality tests, supplemented by our own unit tests using JUnit5 and SpringFramework. An in-memory database allows us to efficiently test CRUD operations.

**Web Page Tests**  
While unit testing has limited use for web pages, we leverage the React Testing Library to test component functionality. Manual integration tests are also performed before deployment to ensure thorough coverage.

**Game Tests**  
Due to the limitations of the LibGDX Framework, we conduct unit tests for static code independent of the LibGDX backend and GPU. Extensive manual integration testing is necessary before each deployment to ensure comprehensive testing.
<img width="474" alt="Bildschirmfoto 2024-05-15 um 13 40 44" src="https://github.com/SE-TINF22B6/Underwatch/assets/122724021/b7ad673e-3639-4ede-bc91-415bf502e642">

### Conclusion

During our testing setup, we learned that unit testing is crucial for static quality analysis. Thus, selecting an easily testable framework is essential for efficient application integration. Unfortunately, our chosen framework hindered this efficiency, necessitating a reliance on integration and manual testing to identify bugs in our codebase.

**Comment by giomskii** [comment]
Hello Team Underwatch,
your test report is quite detailed and covers even potential issues and restrictions of the used test librarys. I think its also good to see that you have already planned ways to evade these problems.
Nonetheless, keep up with the good work! 👍🏼 - Team DHBWhub

**Comment by ghost** [comment]
Dear Underwatch Team, your project structure and testing strategy are impressive! Using SpringBootRest and JUnit5 for backend tests, React Testing Library for web pages, and thorough manual integration testing for the game shows your commitment to quality. Maybe you can also consider adding automated end-to-end testing with tools like Cypress for even better coverage. Great work on ensuring comprehensive testing across your backend and webapp components!

**Comment by Challenger3010** [comment]
It's great to see such a comprehensive update on the testing phases for your project! The detailed breakdown of testing methods across different components is really good. It’s particularly interesting how you adapt testing strategies based on the framework limitations and component needs. Thanks for sharing these valuable insights, and looking forward to more updates on the project's progress!

**Comment by Shiny1708** [comment]
I like your breakdown of the different testing frameworks necessary for the different parts of your project. adressing the limitations and thinking about a way to overcome them is good forward thinking. Although intergrations test are very time intesive they provide great value which you also made clear
